src/misc/rwqueue.cpp

```diff
@@ -74,6 +74,53 @@ void RWQueue<T>::Enqueue(T&& item)
 	has_items.notify_one();
 }
 
+template <typename T>
+void RWQueue<T>::BulkEnqueue(std::vector<T>& from_source, const size_t num_requested)
+{
+	// Check that the source can provide the quantity requested
+	assert(num_requested <= from_source.size());
+
+	// The source iterator will be incremented with each chunk enqueued
+	auto source_start = from_source.begin();
+
+	// The remaining count is decremented until no more items need enqueing
+	auto num_remaining = num_requested;
+
+	while (num_remaining > 0) {
+		// Begin exclusive access to the internal queue
+		std::unique_lock<std::mutex> lock(mutex);
+
+		// Limit the quantity to our capacity
+		const auto min_items = std::max(static_cast<size_t>(1),
+		                                static_cast<size_t>(capacity - queue.size()));
```

Comment:
So if we're at full capacity, `min_items` will be 1? Isn't that a problem?

Reply:
Wouldn't this be simpler and easier to understand? I don't think you need that `max(1, ...)` thing at all because `num_remaining` is guaranteed to be a positive number.

```cpp
const auto free_capacity = static_cast<size_t>(capacity - queue.size()));
const auto num_items = std::min(num_remaining, free_capacity);
```

Reply:
> So if we're at full capacity, min_items will be 1? Isn't that a problem?

It's intended; the best case scenario is if the queue can abosorb then entire vector (free_capacity >= num_requested).

The worst-case is if the queue is full: then we need the mutex predicate's quantity to be at least one element to keep blocking. 

> Wouldn't this be simpler and easier to understand? I don't think you need that max(1, ...) thing at all because num_remaining is guaranteed to be a positive number.

> const auto free_capacity = static_cast<size_t>(capacity - queue.size()));
> const auto num_items = std::min(num_remaining, free_capacity);

If the queue is at capacity, then free-capacity falls to 0, which wins in the std::min() comparison, and num_items == 0 (then we'll spin, which we don't want).

So you're really close! Maybe we can turn it into a std::clamp.

Reply:
Yeah okay, and maybe add a condensed version of the above in a comment because it's really not obvious (it wasn't to me, at least 😄).